### DISCUSSION TITLE
CORS에 대하여

### DISCUSSION THREAD

**Discussion by airman5573** [woowacourse-study/2022-woowahan-bansanghwe#18] [Network]
[교차 출처 리소스 공유 (CORS)](https://developer.mozilla.org/ko/docs/Web/HTTP/CORS)

[CORS는 왜 이렇게 우리를 힘들게 하는걸까?](https://evan-moon.github.io/2020/05/21/about-cors/)

### 추가 질문
- SOP는 무엇일까?
- preflight는 왜 필요한걸까?

## SOP
Same Origin Policy

Origin이 같은 경우에만 요청을 보낼 수 있다.

**요청이란?**

브라우저가 서버에게 ~를 달라고 요청하는것을 의미한다.  
html, css, js파일, 이미지, JSON데이터 등을 요청한다.  

**origin**이 뭐야?

naver에가서 `document.location.origin`을 치면 `https://www.naver.com` 이 나와.  
즉, origin은 domain이지. 다만, 내 생각에는 **출처**라는 의미를 강조하기 위해서 domain이 아닌 origin이라는 용어를 쓰는것 같아.  

**그래서, Same Origin Policy가 뭔데?**
그러니까, `www.naver.com`에서 `www.daum.net`으로 요청을 보낼 수 없다는 의미야.  
참고로, 서버가 아니라 브라우저가 막는거야.  

**이미지는 도메인이 달라도 잘만 가져오던데?**
맞아, [css, js, image, iframe, embeded, video..]등은 도메인이 달라도 잘 가져오지.  

**그럼 SOP가 무슨 의미가 있지?**
xmlhttprequest나, fetch. axios를 쓸때는 도메인이 다르면 요청을 못보내.  

**도메인이 완전히 같아야함? 보통 `www.api.my-site.com`이렇게 앞에 subdomain으로 api를 붙이고는 하잖아**
www.api.my-site.com은 www.my-site.com이랑 다른 사이트로 취급한다네.  
https://en.wikipedia.org/wiki/Same-origin_policy#Origin_determination_rules  테스트를 해본건 아닌데, 위키에 이렇게 나와있긴해.  
위키를 믿자구.  

**잠깐 그 전에, SOP가 왜 필요한거야?**
보안 문제인데, 이런 상황을 생각해 보자.
(1) gmail에 로그인 한다
(2) 해커는 비트코인꽁짜로 준다는 가짜 사이트를 만든다
(3) 해커가 나에게 메일을 보낸다
(4) 욕심많은 병민은 '개꿀!'이라며 메일을 열고 버튼을 클릭한다
(5) 해커가 만든 사이트는 iframe으로 gmail inbox를 불러온다. **SOP가 없으면** 로그인 정보가 cookie에 있으니까 마치 내가 요청한것처럼 데이터를 요청한다. 그리고 또 **SOP가 없으면** origin이 달라도 iframe의 document에 접근해서 javascript로 데이터를 긁어올 수 있다. 해커는 이렇게 긁어온 메일들을 자신의 서버로 전송한다
(6) 나는 이렇게 개인정보가 유출되었다.

이렇게 SOP가 중요해.  

**아니 그러면 어떡해,,나는 날씨 API를 사용하고 싶고, 포켓몬 API를 사용하고 싶은데, SOP때문에 다른 도메인으로 요청을 못보내잖아. 이거 맞아?**
위에서  `[css, js, image, iframe, embeded, video..]등은 도메인이 달라도 잘 가져온다`고 말했었잖아, api요청도 **예외**가 있어.  
그건 바로 C.O.R.S.  

## CORS

Cross Origin Resource Sharing.  

SOP의 예외 조항. 서버로부터 받은 응답 문서(HTTP문서)의 Header부분에 있는 Access-Controll-Allow-Origin필드에 요청을 보낸 domain이 들어 있으면 리소스 공유가 가능하다. 즉, 서버가 허락해야 브라우저와 서버가 통신을 할 수 있다.  

**예외조항?**
원래 SOP때문에 도메인(origin)이 다른 서버로는 요청을 못보내는데, 이러면 인터넷 세상이 너무 삭막해지잖아. 그래서 예외 조항을 만든거야.  

**응답 문서(HTTP문서)의 Header부분은 뭐여?**
```
200 OK
Access-Control-Allow-Origin: *
Connection: Keep-Alive
Content-Encoding: gzip
Content-Type: text/html; charset=utf-8
Date: Mon, 18 Jul 2016 16:06:00 GMT
Etag: "c561c68d0ba92bbeb8b0f612a9199f722e3a621a"
Keep-Alive: timeout=5, max=997
Last-Modified: Mon, 18 Jul 2016 02:36:04 GMT
Server: Apache
Set-Cookie: mykey=myvalue; expires=Mon, 17-Jul-2017 16:06:00 GMT; Max-Age=31449600; Path=/; secure
Transfer-Encoding: chunked
Vary: Cookie, Accept-Encoding
X-Backend-Server: developer2.webapp.scl3.mozilla.com
X-Cache-Info: not cacheable; meta data too large
X-kuma-revision: 1085259
x-frame-options: DENY
```

대충 이렇게 생겼어. 그리고 맨 위에 있는 `Access-Control-Allow-Origin: *` 이 부분이 아까 내가 말한 그 필드야.

**저 별모양은 뭐야?**

어떤 도메인이던 나(서버)에게 요청을 보내도 된다는 의미야.  

**아니, 이런식이면,,CORS가 왜 필요해? 그냥 다 허용하면?**

아 그렇지, 지금의 경우만 보면 맞는 말인데, 대부분의 서버가 `*`이렇게 쓰지는 않아. 어떤 도메인만 허용할지 명시를 해놓지.  
예를들어서 `Access-Control-Allow-Origin: https://normal-website.com`이런식으로 말이지!  

**그런데 이상해! 정말,,아니 저 `Access-Control-Allow-Origin`은 브라우저가 요청을 보내고난 후에 응답값으로 오는거잖아. 그러면 이미 서버에서 처리가 다 끝난거 아니야? 그니까,,완전 뒷북아니야?**

맞아! 사실 브라우저가 요청을 보내고 -> 서버에서 처리한뒤 응답값을 보냈는데 -> Access-Control-Allow-Origin에 현재 사이트 주소가 적혀있지 않으면 정말,,웃기는 일이지. 이미 서버에서 처리는 끝났으니까. 그래서 preflight 과정이 있어!

**preflight? 이게 뭔데?**
[이 영상](https://www.youtube.com/watch?v=OMHH2R5PmDs)을 봐봐. 이게 preflight야. 다시말해서 비행을 하기 전에 준비를 하는거지.  
pre-flight.  

**아니 그래서 그게 뭐냐고**
브라우저가 서버에게 API요청을 보낼때 미리 한번 물어보는 작업이야. 아래와 같은 요청을 먼저 보내.  
```
OPTIONS /resource/foo
Access-Control-Request-Method: DELETE
Access-Control-Request-Headers: origin, x-requested-with
Origin: https://foo.bar.org
```

이런 요청을 보내면 서버가 아래와 같이 응답을 돌려줘  
```
HTTP/1.1 204 No Content
Connection: keep-alive
Access-Control-Allow-Origin: https://foo.bar.org
Access-Control-Allow-Methods: POST, GET, OPTIONS, DELETE
Access-Control-Max-Age: 86400
```
이제 브라우저가 저 `Access-Control-Allow-Origin`를 보고 진짜 요청을 보낼지 말지 결정하는거야.  

**아 그러면 요청 보낼때마다 브라우저는 2번 요청을 보내는거야? preflight한번 + (검사 통과했을때)진짜요청 1번 이렇게?**

맞아! 하지만, 브라우저 개발자들도 이건 너무 비효율적이라고 생각 했는지, 어떤 장치를 마련해놨어.  
`Access-Control-Max-Age`보이지? 저게 바로 그 장치야. 다시말해서, 한번 검사에 통과 했으면 `86400`초 동안은 preflight없이 그냥 요청을 보낸다는거지.  

**오호 그렇구만, 그러면,,,모든 모든 요청이 86400초 동안 preflight없이 가는거야?**
음, 이건 정확히는 모르겠어.  
아마도 같은 요청에 한해서 preflight없이 진짜 요청만 갈듯 싶어.  
예를들어서 my-site/api/notes 요청에서 preflight를 통과하고 `Access-Control-Max-Age`를 받았다고 해서, `my-site/api/people`요청도 preflight없이 가지는 않을듯..? 정확하지는 않아.  

**좋아, 어느정도는 이해를 했어. 혹시 CORS검사 없이도 그냥 요청을 보낼수는 절대 없는거야?**
사실, Simple Request라는게 있어.  
특정 조건을 만족하는 xmlhttprequest는 그냥 CORS검사 없이도 그냥 보낼 수 있어!  
이 조건은 솔직히 만족하기 어려운 조건인데, 한번 읽어보는것도 좋아.  
https://developer.mozilla.org/ko/docs/Web/HTTP/CORS#%EB%8B%A8%EC%88%9C_%EC%9A%94%EC%B2%ADsimple_requests

![image](https://develoger.kr/wp-content/uploads/complete-understand.jpg)

## 참고
- https://evan-moon.github.io/2020/05/21/about-cors/
- https://developer.mozilla.org/ko/docs/Glossary/Preflight_request
- https://developer.mozilla.org/ko/docs/Web/HTTP/CORS

**Comment by InKyoJeong** [comment]
짤이 정말 기엽군요
